sys/sema: allow to use ztimer and/or xtimer

haukepetersen opened the pull request:
### Contribution description
On the winding road to get NimBLE running without including `xtimer` or `periph_timer` in the build, the `sema` and `event` modules are the last ones that pull in these dependencies currently. So I am desperately looking for way to free these modules of `xtimer` deps.

The code in this PR, still very prototypic, illustrates one possible solution to decouple the `sema` module from `xtimer` and at the same time adding `ztimer` support. However, this might not be the most elegant solution, so I am happy to hear any alternative ideas on this task!

Regarding what IMO should be possible:
- per default, build with an untouched API, so no existing code ever breaks
- allow to enable the `ztimer` API in parallel to the existing API -> allows for mixing legacy code with new ztimer code
- allow to disable the `xtimer` support and use the `ztimer` support only -> allows to throw out `xtimer`


### Testing procedure
not in focus, yet.

### Issues/PRs references
none

haukepetersen commented:
Forgot to mention: remodeling the `event` module (`event_wait_timeout()`) presents the same class of problem, so a unified concept that can be applied to both would be needed here.

haukepetersen commented:
Did another take on this, and I think with the API part I am quite happy now. The new take applies the exact same style than we use for the `sys/event` module (see also #15789). The `sema_post_timed()` function(s) are simply only available if the fitting timer backend was compiled in.

The only thing that needs more work is the implementation part. Here we need some smarter approach to prevent code duplication I guess...

fjmolinas commented on sys/sema/sema.c:
```diff
@@ -97,6 +103,63 @@ int _sema_wait(sema_t *sema, int block, uint64_t us)
+            mutex_lock(&sema->mutex);
+        }
+        else {
+            ztimer_now_t start = ztimer_now(ZTIMER_MSEC);
+            block = !ztimer_mutex_lock_timeout(ZTIMER_MSEC, &sema->mutex, timeout);
+            uint32_t elapsed = (uint32_t)(ztimer_now(ZTIMER_MSEC) - start);
```
Can't simply this part be abstracted depending on the timer backend?

fjmolinas replied:
Untested but how about something like:

```diff
diff --git a/sys/sema/sema.c b/sys/sema/sema.c
index 05f61f9380..42e7c9e133 100644
--- a/sys/sema/sema.c
+++ b/sys/sema/sema.c
@@ -29,6 +29,12 @@
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
+#if IS_USED(MODULE_XTIMER)
+typedef uint64_t _sema_timeout_t;
+#elif IS_USED(MODULE_ZTIMER)
+typedef uint32_t _sema_timeout_t;
+#endif
+
 void sema_create(sema_t *sema, unsigned int value)
 {
     assert(sema != NULL);
@@ -49,67 +55,13 @@ void sema_destroy(sema_t *sema)
     mutex_unlock(&sema->mutex);
 }
 
-#if !IS_USED(MODULE_SEMA_NO_XTIMER)
-int _sema_wait(sema_t *sema, int block, uint64_t us)
+#if IS_USED(MODULE_XTIMER) || IS_USED(MODULE_ZTIMER)
+int _sema_wait(sema_t *sema, int block, void* clock, _sema_timeout_t timeout)
 {
     assert(sema != NULL);
-
-    if (sema->state != SEMA_OK) {
-        return -ECANCELED;
-    }
-
-    int did_block = block;
-    unsigned old = irq_disable();
-    while ((sema->value == 0) && block) {
-        irq_restore(old);
-        if (us == 0) {
-            mutex_lock(&sema->mutex);
-        }
-#if IS_USED(MODULE_XTIMER)
-        else {
-            uint64_t start = xtimer_now_usec64();
-            block = !xtimer_mutex_lock_timeout(&sema->mutex, us);
-            uint64_t elapsed = xtimer_now_usec64() - start;
-
-            if (elapsed < us) {
-                us -= elapsed;
-            }
-            else {
-                block = 0;
-            }
-        }
-#endif
-
-        if (sema->state != SEMA_OK) {
-            mutex_unlock(&sema->mutex);
-            return -ECANCELED;
-        }
-
-        old = irq_disable();
-    }
-
-    if (sema->value == 0) {
-        irq_restore(old);
-        return (did_block) ? -ETIMEDOUT : -EAGAIN;
-    }
-
-    unsigned int value = --sema->value;
-    irq_restore(old);
-
-    /* only unlock mutex if it was a blocking operation */
-    if (did_block && value > 0) {
-        mutex_unlock(&sema->mutex);
-    }
-
-    return 0;
-}
-#endif
-
 #if IS_USED(MODULE_ZTIMER)
-int _sema_wait_ztimer(sema_t *sema, int block,
-                      ztimer_clock_t *clock, uint32_t timeout)
-{
-    assert(sema != NULL);
+    assert(clock != NULL);
+#endif
 
     if (sema->state != SEMA_OK) {
         return -ECANCELED;
@@ -123,10 +75,15 @@ int _sema_wait_ztimer(sema_t *sema, int block,
             mutex_lock(&sema->mutex);
         }
         else {
+#if IS_USED(MODULE_XTIMER)
+            uint64_t start = xtimer_now_usec64();
+            block = !xtimer_mutex_lock_timeout(&sema->mutex, timeout);
+            uint64_t elapsed = xtimer_now_usec64() - start;
+#elif IS_USED(MODULE_ZTIMER)
             ztimer_now_t start = ztimer_now(ZTIMER_MSEC);
             block = !ztimer_mutex_lock_timeout(ZTIMER_MSEC, &sema->mutex, timeout);
             uint32_t elapsed = (uint32_t)(ztimer_now(ZTIMER_MSEC) - start);
-
+#endif
             if (elapsed < timeout) {
                 timeout -= elapsed;
             }
@@ -158,8 +115,6 @@ int _sema_wait_ztimer(sema_t *sema, int block,
 
     return 0;
 }
-#endif
-
 
 int sema_post(sema_t *sema)
 {
```

fjmolinas replied:
The headers need quite some adaptation with though.

fjmolinas replied:
Mainly wrap all `time` related functionality in `#if IS_USED(MODULE_XTIMER) ||  IS_USED(MODULE_XTIMER) || defined(DOXYGEN)`, and then add some `@pre ztimer or xtimer must be included` to those function definitions.

haukepetersen replied:
Don't think this works well in cases where both ztimer and xtimer are used simultaneously. Imagine some library uses the `sema_wait_timed()` with an uint64_t timeout value, while some other library uses the `sema_wait_timed_ztimer()` function in the same build. As this should be possible, I think there is no way around keeping both API-paths intact for now.

fjmolinas replied:
Yes you are right, I missed this answer before. Lets keep them separate then.

fjmolinas commented on sys/sema/sema.c:
```diff
@@ -46,6 +49,7 @@ void sema_destroy(sema_t *sema)
     mutex_unlock(&sema->mutex);
 }
 
+#if !IS_USED(MODULE_SEMA_NO_XTIMER)
```
I tried to look at how other of our API's take care of this issue:

- `sock`: timeout functionality will only work if `MODULE_XTIMER` is actually pulled in, but the API has no mention for this
- `mutex`, `msg`: timeout functionality is provided by the different timer abstraction modules and the API specifies that.
- `event`: if the modules are included then the timeout functions will be available, the API itself does not say this unless code is checked and an `ifdef` is seen.

IMO the second is the best approach for `core` to avoid timing dependencies. But since in this case it's a sys module, I think the third approach combined with some better API doc would be the best, for example, a `@pre xtimer/ztimer must be included`. A pseudo module could also be nice but not necessary, I would also rather have it as opt-in instead of opt-out, maybe `sema-timeout`. What do you think?

fjmolinas commented on sys/include/sema.h:
```diff
@@ -179,6 +164,49 @@ static inline int sema_try_wait(sema_t *sema)
+ */
+static inline int sema_wait_timed_ztimer(sema_t *sema,
+                                         ztimer_clock_t *clock,
+                                         uint32_t timeout);
```
```suggestion
                                         uint32_t timeout)
{
         return _sema_wait_ztimer(sema,  (timeout != 0), clock, timeout);
}                               
```

fjmolinas commented on sys/include/sema.h:
```diff
@@ -179,6 +164,49 @@ static inline int sema_try_wait(sema_t *sema)
+ * @return  -ECANCELED, if the semaphore was destroyed.
+ * @return  -EAGAIN,    if the semaphore is not posted (only if timeout = 0)
+ */
+static inline int sema_wait_timed_ztimer(sema_t *sema,
```
How about also adding:

```
static inline int sema_wait_timed(sema_t *sema, uint64_t timeout)
{
    return _sema_wait_ztimer(sema,  (timeout != 0), ZTIMER_USEC, timeout);
}
```

when ztimer is used?

fjmolinas replied:
How the exact call is made should be sorted after factoring a bit the C implementation, but I think having this wrapper to keep the same API would make sense.

haukepetersen replied:
I don't really like it, as this adds quite some complexity regarding resolution of actual implementation based on time backends included in the build, as we allow builds with ztimer and xtimer simultaneously. So I think sticking with a strict separation of `sema_wait_timed() -> xtimer` and `sema_wait_timed_ztimer() -> ztimer` in the API is the cleanest solutiion. If (one day) xtimer is deprecated, we can easily introduce the mapping proposed above for backward compatibility then...

fjmolinas replied:
So can you add that wrapper if timer is used? Also, does it make sense to have calls that use `ztimer` and `xtimer`?

fjmolinas replied:
I had missed https://github.com/RIOT-OS/RIOT/pull/15782#discussion_r604705742, then never mind this comment.

fjmolinas commented:
> Cleaned up the API part slightly. The more I think about this, I come to the conclusion that having two, although in large parts identical, implementations of the `_sema_wait_x()` functions seems to be the cleanest solution for now. At some point we will remove the `xtimer` part anyway, right? Then all we need to do is remove the one code path an map the `sema_wait_timed` to the ztimer call and we are all good.

Sure lets got with it as is, seems like there is actually no taste for `sema` atm no?

fjmolinas reviewed (commented):
@haukepetersen there is a sema test in master now, lets keep it simple and simply use the ztimer version in  one of the wait scenario:

```
diff --git a/tests/sema/Makefile b/tests/sema/Makefile
index ce88d28105..0d0af60982 100644
--- a/tests/sema/Makefile
+++ b/tests/sema/Makefile
@@ -1,5 +1,7 @@
 include ../Makefile.tests_common
 
 USEMODULE += sema
+USEMODULE += xtimer
+USEMODULE += ztimer_usec
 
 include $(RIOTBASE)/Makefile.include
diff --git a/tests/sema/main.c b/tests/sema/main.c
index 295c434103..5f0fdd2fa4 100644
--- a/tests/sema/main.c
+++ b/tests/sema/main.c
@@ -26,6 +26,7 @@
 #include "sema.h"
 #include "thread.h"
 #include "xtimer.h"
+#include  "ztimer.h"
 
 #define TEST_TIME_US 1000
 #define TEST_ITERATIONS 4
@@ -48,7 +49,7 @@ static void *second_thread(void *arg)
         printf("THREAD ERROR: sema_wait_timed()");
         *thread_success = -1;
     }
-    if (sema_wait_timed(&test_sema2, TEST_TIME_US) != 0) {
+    if (sema_wait_timed_ztimer(&test_sema2, ZTIMER_USEC, TEST_TIME_US) != 0) {
         printf("THREAD ERROR: sema_wait_timed()");
         *thread_success = -1;
     }
```

(please rebase to get the test in)

fjmolinas commented on sys/sema/sema.c:
```diff
@@ -97,6 +101,63 @@ int _sema_wait(sema_t *sema, int block, uint64_t us)
+    return 0;
+}
+#endif
+
```
```suggestion
```

fjmolinas commented on sys/include/sema.h:
```diff
@@ -176,8 +191,58 @@ static inline int sema_wait(sema_t *sema)
+#endif
+}
+
+#if IS_USED(MODULE_XTIMER) || defined(DOXYGEN)
+/**
+ * @brief   Wait for a semaphore being posted.
+ *
+ * @pre `(sema != NULL)`
+ *
+ * @param[in]  sema     A semaphore.
+ * @param[in]  timeout  Time in microseconds until the semaphore times out.
+ *                      0 does not wait.
+ *
+ * @return  0 on success
+ * @return  -ETIMEDOUT, if the semaphore times out.
+ * @return  -ECANCELED, if the semaphore was destroyed.
+ * @return  -EAGAIN,    if the semaphore is not posted (only if timeout = 0)
+ */
+static inline int sema_wait_timed(sema_t *sema, uint64_t timeout)
+{
+    return _sema_wait_xtimer(sema, (timeout != 0), timeout);
+}
+#endif
```
```suggestion
#if IS_USED(MODULE_XTIMER) || IS_USED(MODULE_ZITMER) || defined(DOXYGEN)
/**
 * @brief   Wait for a semaphore being posted.
 *
 * @pre `(sema != NULL)`
 *
 * @param[in]  sema     A semaphore.
 * @param[in]  timeout  Time in microseconds until the semaphore times out.
 *                      0 does not wait.
 *
 * @return  0 on success
 * @return  -ETIMEDOUT, if the semaphore times out.
 * @return  -ECANCELED, if the semaphore was destroyed.
 * @return  -EAGAIN,    if the semaphore is not posted (only if timeout = 0)
 */
static inline int sema_wait_timed(sema_t *sema, uint64_t timeout)
{
#if IS_USED(MODULE_ZTIMER)
    return _sema_wait_ztimer(sema,  (timeout != 0), ZTIMER_USEC, timeout);
#else
    return _sema_wait_xtimer(sema, (timeout != 0), timeout);
#endif
}
#endif
```

fjmolinas replied:
This was answered in https://github.com/RIOT-OS/RIOT/pull/15782#discussion_r604705742, not a good idea.

fjmolinas reviewed (approved):
Some final comments.

fjmolinas commented:
> @haukepetersen there is a sema test in master now, lets keep it simple and simply use the ztimer version in one of the wait scenario:

Actually, maybe its better to simply duplicate the test

fjmolinas commented:
> > @haukepetersen there is a sema test in master now, lets keep it simple and simply use the ztimer version in one of the wait scenario:
> 
> Actually, maybe its better to simply duplicate the test

With https://github.com/RIOT-OS/RIOT/pull/15782#discussion_r604705742 in mind I think just the patch in https://github.com/RIOT-OS/RIOT/pull/15782#pullrequestreview-629829483 is enough. So if you remove this newline https://github.com/RIOT-OS/RIOT/pull/15782#discussion_r608520978 then I think this one is OK!

haukepetersen commented:
done:
- added call to `sema_wait_timed_ztimer()` to the test application
- removed the newline
- fixed a dependency problem when building without any active timers

fjmolinas reviewed (approved):
ACK!

haukepetersen commented:
squashed

haukepetersen commented:
All green -> go
